### DISCUSSION TITLE
Python (yes, Python!) integration

### DISCUSSION THREAD

**Discussion by cantao** [quasarframework/quasar#9186] [CLI - Capacitor / Cordova modes] [answered]
Hi!

I'm developing a scientific application using Quasar as the frontend and Python+Flask as the backend, running locally on the same machine. Python was chosen because of the SciPy ecosystem.

Everything runs smoothly on Linux or Windows: the frontend passes data to the backend, numbers are crunched, results fly back and we have nice tables and plots.

I can do the same on Android with Cordova. The point in question is that I have to install Termux and manually run the Flask server. Cumbersome, from the point of view of regular (non-developer) user.

I am at a loss here, as how to integrate Quasar and Python+Flask backend in a single APK (if that’s possible at all) and run the Python code as soon as the application starts (Android service, maybe?).

Any hints or tips will be very helpful!

Thanks a lot!

**Comment by TobyMosque** [comment]
the Flask API would be hosted in a webserver.

**Reply by cantao** [reply]
Thanks for the response!

Flask runs locally (127.0.0.1) acting as a bridge between Quasar/Capacitor and the numerical kernel implemented in Python. The application should be able to run without an Internet connection.

Quasar is responsible for file selection, that is then POSTed to Flask, read, processed (a few statistical models) and then the results are sent back to Quasar to display (tables, graphs, etc).

Everything works like a charm on desktop/electron, because is easy to start a script to run the numerical kernel.

Unfortunatelly there are no (to my knowledge) JavaScript libraries as powerful as the Python Scientific Stack :-(

**Reply by ontwikkelfabriek** [reply]
You could take a look at these repo's that run Python code in the browser ( so you can use it in combination with Quasar ). Although you probably have to ditch Flask and run 'pure' python scripts:

https://github.com/qquick/Transcrypt
https://github.com/brython-dev/brython

Also a native Android  solution to run Python on Android:( native so no Quasar....)
https://www.qpython.org/

There's even a cross platform IOS and Android way:( native so no Quasar....)
https://kivy.org/#home

**Reply by cantao** [reply]
Hi @ontwikkelfabriek , thanks for the directions!

Transcrypt and Brython are amazing projects, but unfortunatelly neither can run Numpy and Scipy.

I'm looking into QPython and Kivy has a "service" mode that looks promising! Maybe I can start a background service and establish some kind of socket communication (if that's possible, zero experience with Android).

I'm also playing with Pyodide (https://pyodide.org/), lets see what happens.

**Reply by ontwikkelfabriek** [reply]
@cantao 

Ah that Pyodide is new to me. Talking about WASM,
here's a thread about using a WASM in Quasar:
https://forum.quasar-framework.org/topic/7123/using-ffmpeg-wasm-in-quasar

**Comment by hawkeye64** [comment, marked as answer]
@cantao Open this link, then look for "afterDev"
https://next.quasar.dev/quasar-cli/quasar-conf-js#property-build

You can use this to start a server just before quasar cli opens a web browser for you.

**Reply by cantao** [reply]
Hi @hawkeye64 , thanks for the link! I'm still on Quasar v1, trying to use boot files,. Lets see what I can come up with.

**Reply by cantao** [reply]
@hawkeye64 Amazing! Thanks for the link!

**Reply by hawkeye64** [reply]
@cantao Can you tag as answered. Thanks.

**Reply by cantao** [reply]
Sure, thanks!

**Comment by mohamed99elsokary** [comment]
Did you get it working ?

**Reply by cantao** [reply]
I am sorry, but no. At some point I simply gave up and rewrote the numerical kernel in Javascript. Working with Python *and* Quasar at the same time on Android turned into a huge pile of questionable hacks.
